doc/NAC/PnacSonicHLD.md

```diff
@@ -0,0 +1,374 @@
+***Cons:***
+Currently SONiC distribution does not support STP/RSTP protocol.
+Port state should be set taking into account both NAC status on a port and RSTP/MSTP status. 
+In future, when RSTP/STP code gets implemented the port state need to be controlled using combination of NAC and RSTP/STP protocol. This is complicates the solution.
+### 3.1.2 Option-2: Allowing or Blocking user traffic by configuring L2-ACL filters 
+In this design option, software shall use two ACL filters,
+
+L2-ACL filter to allow EAPOL traffic
+L2-ACL filter to block the data traffic
+
+When a bridge port is in unauthorized state following filters shall be configured (index indicates priority of the filter).
+
+L2-ACL filter to allow EAPOL traffic
+L2-ACL filter to block the data traffic
+
+When a bridge port is in authorised state filters then following filter is configured.
+
+L2-ACL filter to allow EAPOL traffic
+
+
+***Pros:***
+Allowing and blocking traffic is achieved by L2-ACL filter using existing SAI APIs. Hence, no change in SAI API is required to achieve the functionality.  
+***Cons:***
+Total number of L2-ACL filters resources are limited for most of forwarding ASIC.
+Hence consuming 4 filters per interface will consume large number entries L2-ACL filter table, if NAC is enabled for interfaces. This will reduce useful number entries in L2-ACL filter table.
+### 3.1.3 Option-3: Allowing or Blocking user traffic by configuring learn mode
+In this design option, based on port Authorisation method, “learn_mode” for a port is set one of the following modes,
+
+“SAI_BRIDGE_PORT_FDB_LEARNING_MODE_DROP”
+ “SAI_BRIDGE_PORT_FDB_LEARNING_MODE_HW”
+
+When port is in unauthorized state after enabling NAC feature then port learn mode is configured as “SAI_BRIDGE_PORT_FDB_LEARNING_MODE_DROP”.
+
+When port is in authorized state after enabling NAC feature then port learn mode is configured as  “SAI_BRIDGE_PORT_FDB_LEARNING_MODE_HW”
+
+
+***Pros:***
+No change in SAI APIs is required to implement this design option.
+In both learn_modes EAPOL traffic is punted to CPU i.e. no specific configuration is required for punting EAPOL traffic.
+No consumption L2-Filter entries.
+
+***Cons:***
+None
+
+After evaluating each of the options and their Pro/Cons, Option-3 is the priority-1 option and hence chosen as the option for PNAC design.
+
+## 3.2 PNAC software architecture
+Following figure explains how various components are based on microservices architecture. SONiC natively uses docker containers to partition various SW components based on grouping of functionalities. 
+![PNAC Use Case](images/PNAC_SW_Architecture.png)
+
+A new Docker container is introduced that implements control plane of 802.1x standard with Authenticator functionality only. Within the "hostapd" container a new layer “hostapd_wrapper” is added. "hostapd_wrapper" enables communication between "hostapd" and other SONiC components.
```

Comment:
A new docker is not recommended. "hostapd" already exists in the "macsec" docker of SONiC. Maintaining duplicate codebases is an issue. 

Also 802.1x Authenticator (hostapd) will be needed for Host-to-Switch MACsec, whenever it is implemented in SONiC. That will preferably use the "macsec" docker's hostapd and not this new docker's hostapd, since Switch-to-Switch MACsec is implemented in the "macsec" docker.

Reply:
If we are using a common container for both features i.e. PNAC and macsec, we are creating lot of tight coupling between these features. Deploying selective features shall be difficult. Hence it was decided to have a sperate container for PNAC (considering a small footprint of hostapd. Also any new nac features additions will be simpler and easier for maintenance.